woocommerce/google-listings-and-ads
# Add automatic settings for shipping rates

### Changes proposed in this Pull Request:



Implements part of #1117.

This PR adds a new `automatic` setting to the shipping rates (we currently have `manual` and `flat`). This new setting will instruct the backend to automatically sync the Merchant Center shipping settings with WooCommerce shipping settings. It uses the WooCommerce shipping zones to get a good estimate of shipping rates in the defined countries and submits them to Google.

I say "estimate" because the shipping zones can be defined in so many complex ways and the Google shipping settings can only handle a limited scope of what can be defined in Woo. We try to map most of the settings 1:1 but there are many cases where we either have to skip or pick one of many available options because there can't be a 1:1 mapping. For example, defining a shipping rate based on a formula containing variables is not supported in Merchant Center; therefore, we skip those shipping rates at the moment.

I had to do some refactoring to the `MerchantCenterService` and extract a `TargetAudience` service from it because the functionality related to fetching the target countries was required in the `Settings` class and since this class is also used in `MerchantCenterService`, if I had used the methods from `MerchantCenterService` it would've created a dependency loop. Most of the changes in this PR are side effects of refactoring this class.

### Detailed test instructions:


1. Define multiple shipping zones each containing various regions and shipping methods and rates for shipping classes.
2. Send a `POST` request to `wc/gla/mc/settings` and include the following data as JSON body: `{ "shipping_rate":"automatic" }`
3. Confirm that the setting is updated in DB. It's stored in `wp_options` under key: `gla_merchant_center`
4. Send a `POST` request to `wc/gla/mc/settings/sync`
5. Confirm that it's successful
6. Check the shipping rates in your MC account and confirm that they match what you defined in Woo shipping zones


### Additional details:


### Changelog entry

>
